### DISCUSSION TITLE
Socket, TCP, HTTP, HttpClient and constant traffic drops in opposit to Python using asyncio and aiohttp

### DISCUSSION THREAD

**Discussion by okarpov** [dotnet/runtime#87572] [General]
Hello everyone,

I have faced with very strange situation, which causes constant outgoing traffic drops after some time of running .net console app.
I do use single instance of HttpClient and do send GET requests to different servers constantly using SendAsync + HttpRequestMessage. But after some time i see that it starts getting constant traffic drops.

Any idea what may cause it? Is there any specific cases/conditions when you able to get such/similar issue?

similar Python script using asyncio and aiohttp produces smooth constant traffic without jumps and drops.

Thank you for any idea, explenation, details, observation.

![image](https://github.com/dotnet/runtime/assets/2629775/8e5a4014-8102-43fb-8883-8a3ec281026c)

**Comment by frankhaugen** [comment]
This is a bit light on details, so I'm just throwing something out there based on my "gut": Garbage Collection

If there is any behaviour that starts after a while and can be described as "pulsating" I'm looking into when GC is happening as a first step. You might be doing some IO or HttpClient anti-patterns that cause very aggressive GC behavour.

**Reply by okarpov** [reply]
well, because of limited resources we do GC from time to time or at some few conditions

**Reply by frankhaugen** [reply]
Manual GC is very much not reccomended, and the runtime will run GC at it's own discetion. If you are in a situation where you MUST run GC, then I would say that you have some memory/data anti-patterns, escpecially writing code that is hard for the runtime/jit to recognize "collollectable garbage" from. 

You should look in your code to see if there are missing using-blocks, (there are analyzers built-in in dotnet for that)

**Reply by okarpov** [reply]
there are files wrapped into using and static HttpClient with undergoing using blocks for HttpRequest/Response and http streams

**Reply by okarpov** [reply]
"(there are analyzers built-in in dotnet for that)" could you, please, point out?

**Reply by frankhaugen** [reply]
> "(there are analyzers built-in in dotnet for that)" could you, please, point out?
Sorry @okarpov, it turns out to be an IDE -feature of Rider

I'm not able to give you much more advice, Good luck!
